modules/proxy/mod_proxy_http.c

```diff
@@ -1615,6 +1649,62 @@ int ap_proxy_http_process_response(proxy_http_req_t *req)
             do_100_continue = 0;
         }
 
+        if (upgrade) {
+            apr_status_t rv;
+            proxy_tunnel_rec *tunnel;
+            apr_interval_time_t client_timeout = -1,
+                                backend_timeout = -1;
+
+            /* If we didn't send the full body yet, do it now */
+            if (do_100_continue) {
+                req->expecting_100 = 0;
+                status = send_continue_body(req);
+                if (status != OK) {
+                    return status;
+                }
+            }
+
+            ap_log_rerror(APLOG_MARK, APLOG_DEBUG, 0, r, APLOGNO()
+                          "HTTP: tunneling protocol %s", upgrade);
+
+            rv = ap_proxy_tunnel_create(&tunnel, r, origin, "HTTP");
+            if (rv != APR_SUCCESS) {
+                ap_log_rerror(APLOG_MARK, APLOG_ERR, rv, r, APLOGNO()
+                              "can't create tunnel for %s", upgrade);
+                return HTTP_INTERNAL_SERVER_ERROR;
+            }
+
+            /* Set timeout to the lowest configured for client or backend */
+            apr_socket_timeout_get(backend->sock, &backend_timeout);
+            apr_socket_timeout_get(ap_get_conn_socket(c), &client_timeout);
+            if (backend_timeout >= 0 && backend_timeout < client_timeout) {
+                tunnel->timeout = backend_timeout;
+            }
+            else {
+                tunnel->timeout = client_timeout;
+            }
+
```

Comment:
Why do we need to align the timeouts and and have a procedure different from mod_proxy_wstunnel. This can be confusing.

Reply:
In mod_proxy_wstunnel there is also a single (idle_)timeout for the tunneling part, this is how the ap_proxy_tunnel loop works (I don't it makes sense to have two separate timeouts for client and backend once it's not HTTP anymore).

Whether we should use the min, max, always use the backend's, or have a dedicated idle timeout is an open question, though. I'd say the backend's, maybe?

Reply:
I saw that a single timeout is used in `ap_proxy_tunnel` and this is not the point I wanted to tackle. I was more worried that people get confused timeout wise with the setting done by `ProxyWebsocketIdleTimeout` which is `mod_proxy_wstunnel` related.  But this is not backported yet and probably before it is backported / 2.5[6] is released we have all that in `mod_proxy_http` such there is no need for having `ProxyWebsocketIdleTimeout`  ever in a stable release in its current form.
Once we have all that in `mod_proxy_http` we can think of how to handle it with the same parameters (approach) we currently use for timeouts. I think we already use two timeouts in `mod_proxy_wstunnel`in trunk. One for idle timeouts and one (or two different ones for client and server) for blocking timeouts.

Reply:
The plan would indeed be to deprecate proxy_wstunnel and have the upgrades handled by proxy_http, so `ProxyWebsocketIdleTimeout` would become ProxyAsyncTimeout and for the HTTP protocol we'd keep the usual `Timeout` (client side) and `ProxyTimeout` / worker `timeout=` on the origin side (which are the ones used by proxy_wstunnel too until the upgrade).

I'm on the async side of things now for this to possible ;)
Though as you said this is trunk only today in proxy_wstunnel.